devito/passes/clusters/aliases.py

```diff
@@ -81,244 +81,303 @@ def cire(clusters, mode, sregistry, options, platform):
+
+                mapper.add(i, make, terms)
+
+        return mapper
 
-    def _nrepeats(self, cluster):
+    def _generate(self, exprs, exclude):
+        """
+        Generate one or more extractions from ``exprs``. An extraction is a
+        set of CIRE candidates which may be turned into aliases. Two different
+        extractions may contain overlapping sub-expressions and, therefore,
+        should be processed and evaluated indipendently. An extraction won't
+        contain any of the symbols appearing in ``exclude``.
+        """
         raise NotImplementedError
 
-    def _extract(self, exprs, context, n):
+    def _lookup_key(self, c):
+        """
+        Create a key for the given Cluster. Clusters with same key may be
+        processed together in the search for CIRE candidates. Clusters should
+        have a different key if they must not be processed together, e.g.,
+        when this would lead to violation of data dependencies.
+        """
         raise NotImplementedError
 
-    def _in_writeto(self, dim, cluster):
+    def _betterv(self, delta_flops, delta_ws):
+        """
+        Given the deltas in flop reduction and working set size increase of two
+        Variants, return True if the second variant is estimated to deliever
+        better performance, False otherwise.
+        """
         raise NotImplementedError
 
-    def _selector(self, e, naliases):
+    def process(self, clusters):
         raise NotImplementedError
 
 
-class CireInvariants(Cire):
+class CireInvariants(CireTransformer, Queue):
 
     optname = 'invariants'
 
-    def _nrepeats(self, cluster):
-        return 1
+    def __init__(self, sregistry, options, platform):
+        super().__init__(sregistry, options, platform)
 
-    def _rule(self, e):
-        return (e.is_Function or
-                (e.is_Pow and e.exp.is_Number and e.exp < 1))
+        self.opt_maxpar = True
+        self.opt_schedule_strategy = None
 
-    def _extract(self, exprs, context, n):
-        mapper = Uxmapper()
-        for prefix, clusters in context.items():
-            if not prefix:
-                continue
+    def process(self, clusters):
+        return self._process_fatd(clusters, 1, xtracted=[])
 
-            exclude = set().union(*[c.scope.writes for c in clusters])
-            exclude.add(prefix[-1].dim)
+    def callback(self, clusters, prefix, xtracted=None):
+        if not prefix:
+            return clusters
+        d = prefix[-1].dim
 
-            for e in exprs:
-                for i in search(e, self._rule, 'all', 'bfs_first_hit'):
-                    if {a.function for a in i.free_symbols} & exclude:
-                        continue
-                    mapper.add(i, self._make_symbol)
+        # Rule out extractions that would break data dependencies
+        exclude = set().union(*[c.scope.writes for c in clusters])
 
-        return mapper
+        # Rule out extractions that are *not* independent of the Dimension
```

Comment:
are not independent \equivalent to are dependent (?)

Reply:
indeed, will fix